### PR TITLE
Reduce code repetition in `datafusion/functions` mod files

### LINKED ISSUE #10397: Reduce repetition in datafusion::functions using macros
I think it is possible to take `Vec` like what functions-array macro does

```rust
($UDF:ty, $EXPR_FN:ident, $DOC:expr , $SCALAR_UDF_FN:ident) => {
        paste::paste! {
            // "fluent expr_fn" style function
            #[doc = $DOC]
            pub fn $EXPR_FN(arg: Vec<Expr>) -> Expr {
                Expr::ScalarFunction(ScalarFunction::new_udf(
                    $SCALAR_UDF_FN(),
                    arg,
                ))
            }
```

_Originally posted by @jayzhan211 in https://github.com/apache/datafusion/pull/10321#discussion_r1585700474_

### PR DESCRIPTION
**Pull request by MohamedAbdeen21**
## Which issue does this PR close?

Possibly closes #10397.

## Rationale for this change

Extending the `export_function!` macro to be used in math and core functions modules, reducing code repetition.

Some functions expect a single `Expr` arg and others expect a single `Vec<Expr>` args. I took a page from Python's 1-tuples (where `(1)` is evaluated to `int(1)`, but `(1,)` is evaluated to `tuple(1)` and now single args followed by a comma are treated as a `Vec<Expr>` but single args without trailing commas are treated as `Expr`.

Variadic functions are unchanged `arg1 arg2 ...`, but should not contain commas.

I left a comment regarding the special case of `get_field`.

## What changes are included in this PR?

Extending the `export_function!` macro to accept both variadic functions and functions that take single `Vec<Expr>` argument. Uses the rust forums link made by @jayzhan211 https://users.rust-lang.org/t/macro-repetition-with-multiple-rules/110816/2?u=jayzhan

## Are these changes tested?

Passes existing tests

## Are there any user-facing changes?

No

### CODE REVIEW

File: datafusion/functions/src/macros.rs [hunk 1]
```diff
@@ -59,6 +59,30 @@ macro_rules! export_functions {
...
+    };
+    // variadic arguments (zero or more arguments, without commas)
+    ($FUNC:ident, $DOC:expr, $($arg:ident)*) => {
+        #[doc = $DOC]
+        pub fn $FUNC($($arg: datafusion_expr::Expr),*) -> datafusion_expr::Expr {
+            super::$FUNC().call(vec![$($arg),*])
+        }
+    };
+    // special to get_field function in datafusion/functions/src/core/mod.rs
+    ($FUNC:ident, $DOC:expr, $arg:ident, $arg2:ident) => {
```
Comment (MohamedAbdeen21) [inline datafusion/functions/src/macros.rs:78]: This works fine, but I hate having special conditions in shared functions/macros. Only alternative I can currently think of is rewriting `get_field` and make both args of type `Expr`

File: datafusion/functions/src/core/mod.rs [hunk 2]
```diff
@@ -41,70 +38,40 @@ make_udf_function!(r#struct::StructFunc, STRUCT, r#struct);
...
+    "Returns a struct with the given names and arguments pairs",
+    args,
+),(
+    get_field,
+    "Returns the value of the field with the given name from the struct",
+    arg1, field_name
+),(
+    coalesce,
+    "Returns `coalesce(args...)`, which evaluates to the value of the first expr which is not NULL",
+    args,
```
Comment (MohamedAbdeen21) [inline datafusion/functions/src/core/mod.rs:76]: Single arguments followed by a single comma indicate a vector argument. The only downside here is that rust-analyzer is unable to identify the type of `args` and the dev is expected to read the doc of `export_functions!` to notice that `args` is a vector and not an Expr.

File: datafusion/functions/src/macros.rs (hunk 1 above)
Comment (jayzhan211) [reply datafusion/functions/src/macros.rs:78]: I don't agree to rewrite the second argument to Expr.
I think we can remove it out of macro and define it specially given it's specialty

Comment (MohamedAbdeen21) [reply datafusion/functions/src/macros.rs:78]: Thought about that, the only problem is how can we split the definition of `pub fn functions()`  between two different macros, without pulling it into it's own macro?

Comment (jayzhan211) [reply datafusion/functions/src/macros.rs:78]: Maybe move `functions` out of macro?

File: datafusion/functions/src/macros.rs [hunk 3]
```diff
@@ -36,25 +36,31 @@
...
-                #[doc = $DOC]
-                /// Return $name(arg)
-                pub fn $FUNC($($arg: datafusion_expr::Expr),*) -> datafusion_expr::Expr {
-                    super::$FUNC().call(vec![$($arg),*],)
-                }
-            )*
+    ($(($FUNC:ident, $DOC:expr, $($arg:tt)*)),*) => {
+        $(
+            // switch to single-function cases below
+            export_functions!(single $FUNC, $DOC, $($arg)*);
```
Comment (MohamedAbdeen21) [inline datafusion/functions/src/macros.rs:47]: A little hack to avoid making a separate macro.

File: datafusion/functions/src/unicode/mod.rs [hunk 4]
```diff
@@ -47,22 +47,64 @@ make_udf_function!(reverse::ReverseFunc, REVERSE, reverse);
...
     pub fn char_length(string: Expr) -> Expr {
         character_length(string)
     }
 
-    #[doc = "the number of characters in the `string`"]
-    pub fn character_length(string: Expr) -> Expr {
-        super::character_length().call(vec![string])
-    }
-
     #[doc = "Returns a value in the range of 1 to N if the string str is in the string list strlist consisting of N substrings"]
```
Comment (alamb) [inline datafusion/functions/src/unicode/mod.rs:108]: why not use the macros for these function too?

Comment (MohamedAbdeen21) [reply datafusion/functions/src/unicode/mod.rs:108]: Oh wow I missed the `find_in_set`. The others are aliases, so the macro doesn't apply

Comment (MohamedAbdeen21) [reply datafusion/functions/src/unicode/mod.rs]: @alamb extremely sorry about the accidental force push, meant to push with lease after the rebase ...

File: datafusion/functions/src/string/mod.rs [hunk 5]
```diff
@@ -70,117 +70,98 @@ make_udf_function!(uuid::UuidFunc, UUID, uuid);
...
+        "Converts a string to uppercase.",
+        arg1
+    ),(
+        uuid,
+        "returns uuid v4 as a string value",
+    ));
 
     #[doc = "Removes all characters, spaces by default, from both sides of a string"]
-    pub fn btrim(args: Vec<Expr>) -> Expr {
+    pub fn trim(args: Vec<Expr>) -> Expr {
```
Comment (jayzhan211) [inline datafusion/functions/src/string/mod.rs:155]: Is this rename intended?

Comment (jayzhan211) [reply datafusion/functions/src/string/mod.rs:155]: Oh, it is alias

**Review by jayzhan211 — APPROVED** [review]
👍

File: datafusion/functions/src/macros.rs [hunk 6]
```diff
@@ -36,25 +36,31 @@
...
+        )*
+    };
+
+    // single vector argument (a single argument followed by a comma)
```
Comment (alamb) [inline datafusion/functions/src/macros.rs:51]: that is pretty fancy 👍

**Review by alamb — COMMENTED** [review]
Thanks @jayzhan211  and @MohamedAbdeen21

**Comment by MohamedAbdeen21** [comment]
Thanks for the review! @jayzhan211 @alamb
